src/Processors/PingPongProcessor.h

```diff
@@ -0,0 +1,104 @@
+#pragma once
+
+#include <Processors/IProcessor.h>
+#include <base/unit.h>
+#include <Processors/Chunk.h>
+#include <Common/logger_useful.h>
+
+namespace DB
+{
+
+/*
+ * Processor with N inputs and N outputs. Moves data from i-th input to i-th output as is.
+ * It has a pair of auxiliary ports to notify another instance by sending empty chunk after some condition holds.
+ * You should use this processor in pair of instances and connect auxiliary ports crosswise.
+ *
+ *     ╭─┴───┴───┴───┴───┴─╮       ╭─┴───┴───┴───┴───┴─╮
+ *     │                   ├─ aux ⟶│                   │
+ *     │ PingPongProcessor │       │ PingPongProcessor │
+ *     │                   │⟵ aux ─┤                   │
+ *     ╰─┬───┬───┬───┬───┬─╯       ╰─┬───┬───┬───┬───┬─╯
+ *
+ * One of the processors starts processing data, and another waits for notification.
+ * When `isReady` returns true, the first stops processing, sends a ping to another and waits for notification.
+ * After that, the second one also processes data until `isReady`, then send a notification back to the first one.
+ * After this roundtrip, processors bypass data from regular inputs to outputs.
```

Comment:
The executor should be aware of these dependencies to choose the proper transform to execute. Without cross connection, we won't be able to understand that we need to run transform from the left side to continue right, and the right would be stuck.

Reply:
seems like it may be reformulated as "executor should not try to continue executing right transform until some async event happened (sync primitive will be available)". we have [some support](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/IProcessor.h#L155) for async events on pipeline level. but from the first glance all the code is written only for Linux. probably it is not a good idea to try to rebuilt `PingPongProcessor` logic on top of it.